src/server/qgsconfigcache.h

```diff
@@ -65,21 +103,152 @@ class SERVER_EXPORT QgsConfigCache : public QObject
 
     QCache<QString, QDomDocument> mXmlDocumentCache;
-    QCache<QString, QgsProject> mProjectCache;
+    QCache<QString, std::pair<QDateTime, std::unique_ptr<QgsProject> > > mProjectCache;
 
-  private slots:
-    //! Removes changed entry from this cache
+    std::unique_ptr<QgsAbstractCacheStrategy> mStrategy;
+
+  private:
+    //! Insert project in cache
+    void cacheProject( const QString &path, QgsProject *project );
+
+    static QgsConfigCache *sInstance;
+
+  public slots:
+    //! Remove entry
     void removeChangedEntry( const QString &path );
+
+    //! Remove all cached entries
+    void removeChangedEntries();
 };
 
+
+#ifndef SIP_RUN
+
+/**
+ * \ingroup server
+ * \brief File system cache strategy for server configuration
+ * \since QGIS 3.24
+ */
+class SERVER_EXPORT QgsFileSystemCacheStrategy : public QgsAbstractCacheStrategy
+{
+  public:
+    QgsFileSystemCacheStrategy();
+
+    //! The name of the strategy
+    QString name() const override { return QStringLiteral( "filesystem" ); };
+
+    //! Attach cache to this strategy
+    void attach( QgsConfigCache *cache ) override;
+
+    /**
+     * Called when an entry is removed from cache.
+     * \param path The path of the project
+     */
+    void entryRemoved( const QString &path ) override;
+
+    //! Called when an entry is inserted
+    void entryInserted( const QString &path ) override;
+
+  private:
+    //! For invalidating files
+    QFileSystemWatcher mFileSystemWatcher;
+};
+
+
+/**
+ * \ingroup server
+ * \brief Periodic system cache strategy for server configuration
+ * \since QGIS 3.24
+ */
+class SERVER_EXPORT QgsPeriodicCacheStrategy: public QgsAbstractCacheStrategy
+{
+  public:
+    QgsPeriodicCacheStrategy( int interval = 3000 );
+
+    //! The name of the strategy
+    QString name() const override { return QStringLiteral( "periodic" ); };
+
+    /**
+     * Set the invalidation check interval for PeriodicStrategy
+     *
+     * If \a msec is set to a positive value then the cache will check for invalidation
+     * every \a msec milliseconds
+     *
+     * \since QGIS 3.24
+     */
+    void setCheckInterval( int msec );
+
+    //! Return the invalidation check interval
+    int checkInterval() const { return mInterval; }
+
+    //! Attach cache to this strategy
+    void attach( QgsConfigCache *owner ) override;
+
+    /**
+     * Called when an entry is removed from cache.
+     * \param path The path of the project
+     */
+    void entryRemoved( const QString &path ) override;
+
+    //! Called when an entry is inserted
+    void entryInserted( const QString &path ) override;
+
+  private:
+    //! Timer for checking cache invalidation
+    QTimer mTimer;
+
+    //! Interval in ms between two invalidation check
+    int mInterval;
+};
+
+/**
+ * \ingroup server
+ * \brief Null system cache strategy for server configuration, it basically does nothing,
+ * completely disabling cache invalidation
+ * invalidation
```

Comment:
Doc looks strange here.

Reply:
> Doc looks strange here.

In what matter ? It describes what the strategy does, i.e nothing (may be `it basically does nothing` is useless)

Reply:
Changed wording.